### PR TITLE
Bluetooth: has: Add support for Read Presets Request operation

### PR DESCRIPTION
**Pull request by MariuszSkamra**
This extends client implementation with Read Presets Request operation
support. The shell and bsim test has been extended with this functionality as well.

### CODE REVIEW

**Comment by MariuszSkamra** [comment]
@szymon-czapracki FYI

File: subsys/bluetooth/audio/has_client.c
Comment (Thalley) [inline subsys/bluetooth/audio/has_client.c:211]:
```suggestion
			const struct bt_has_preset_delete_val value = {
```

File: subsys/bluetooth/audio/has_client.c
Comment (Thalley) [inline subsys/bluetooth/audio/has_client.c:299]:
```suggestion
	(void)memset(req, 0, sizeof(*req));
```

File: subsys/bluetooth/shell/has_client.c
Comment (Thalley) [inline subsys/bluetooth/shell/has_client.c:115]: Any particular reason why `index` is hex and `count` is decimal?

File: subsys/bluetooth/shell/has_client.c
Comment (Thalley) [inline subsys/bluetooth/shell/has_client.c:129]: Suggest to replace this with `inst` - i.e. verify that `inst` is not NULL. If `inst` is not NULL, that means that discovery has completed and that the `ctx_shell` is also set.

File: tests/bluetooth/bsim_bt/bsim_test_audio/src/has_client_test.c
Comment (Thalley) [inline tests/bluetooth/bsim_bt/bsim_test_audio/src/has_client_test.c:60]: `err` is always `0` here - Does it make sense to "log" it?

File: tests/bluetooth/bsim_bt/bsim_test_audio/src/has_client_test.c
Comment (Thalley) [inline tests/bluetooth/bsim_bt/bsim_test_audio/src/has_client_test.c:71]:
```suggestion
	const bool result = value->properties == expected_properties &&
			    !strcmp(value->name, expected_name);
```

File: tests/bluetooth/bsim_bt/bsim_test_audio/src/has_client_test.c
Comment (Thalley) [inline tests/bluetooth/bsim_bt/bsim_test_audio/src/has_client_test.c:73]:
```suggestion
	if (!result) {
```

File: tests/bluetooth/bsim_bt/bsim_test_audio/src/has_client_test.c
Comment (Thalley) [inline tests/bluetooth/bsim_bt/bsim_test_audio/src/has_client_test.c:102]: What's the reasoning for setting a failing flag and then `FAIL` on it later (or even before in the last case here), instead of just calling `FAIL`?

**Review by Thalley — CHANGES_REQUESTED** [review]

File: tests/bluetooth/bsim_bt/bsim_test_audio/src/has_client_test.c
Comment (MariuszSkamra) [reply tests/bluetooth/bsim_bt/bsim_test_audio/src/has_client_test.c:102]: Yes, it makes no sense at all.

File: subsys/bluetooth/shell/has_client.c
Comment (MariuszSkamra) [reply subsys/bluetooth/shell/has_client.c:115]: I've already started using hex for `index` in some other place, thus did the same here. Something tells me that `count` should be in decimal

File: subsys/bluetooth/audio/has_client.c
Comment (MariuszSkamra) [reply subsys/bluetooth/audio/has_client.c:121]: I guess `properties` shall not be verified, because The specification may change so that there will be more property flags defined.
Regarding `is_last` I would not strict to the 0 or 1 even if specification defines those values only. I guess it's sane to provide the preset to the application anyway even if the `is_last` parameter is invalid.

But I guess it would be nice to have at least warning log here.

File: subsys/bluetooth/audio/has_client.c
Comment (MariuszSkamra) [reply subsys/bluetooth/audio/has_client.c:185]: You do not know. the `Previous Index` gives clue. For example if you as a client have e.g. presets in your cache like:
Index: 1, Index: 10, Index: 33
and you receive a Generic Update: Index: 11, Index Previous: 0
It tells you to remove indices 1 and 10.

File: include/zephyr/bluetooth/audio/has.h
Comment (MariuszSkamra) [reply include/zephyr/bluetooth/audio/has.h:194]: It gets 5 callbacks. The last one have `is_last` flag set

File: include/zephyr/bluetooth/audio/has.h
Comment (MariuszSkamra) [reply include/zephyr/bluetooth/audio/has.h:147]: Yes, I have considered. I decided to have one because, This one callback would be used to handle preset list updates. Moreover `Generic Update` notification and `Read Preset Response` are so similar I wanted to have one, common callback for those

File: subsys/bluetooth/audio/has_client.c
Comment (MariuszSkamra) [reply subsys/bluetooth/audio/has_client.c:65]: You know, I'll remove this. I thought it would be nice to handle multiple requests to make use of EATT as well as be able to set the active preset when Read Preset operation is in progress, but to reduce the memory consumption, tel's handle one request at the time

File: include/zephyr/bluetooth/audio/has.h
Comment (Thalley) [reply include/zephyr/bluetooth/audio/has.h:184]: OK, so it is _up to_ count? 

I think that should be specified somehow, e.g. either with stating `up to @p count` or perhaps rename `count` to `max_count`?

File: subsys/bluetooth/audio/has_client.c
Comment (Thalley) [reply subsys/bluetooth/audio/has_client.c:113]: If the core specifies that `is_last` is either 0 or 1, and the server replies with a value other than that, then it's a non-conformant server, and it's pretty risky to consider the remainder of the response as valid IMO. 

Is there a BI test case for it?

File: include/zephyr/bluetooth/audio/has.h
Comment (MariuszSkamra) [reply include/zephyr/bluetooth/audio/has.h]: Done :)

File: include/zephyr/bluetooth/audio/has.h
Comment (MariuszSkamra) [reply include/zephyr/bluetooth/audio/has.h]: Removed. The `user_data` was there because I wanted to provide some way to keep the operation context (I removed the multiple requests)

File: subsys/bluetooth/audio/has_client.c
Comment (MariuszSkamra) [reply subsys/bluetooth/audio/has_client.c:113]: No, there is no such test case

Comment (Thalley) [reply subsys/bluetooth/audio/has_client.c:113]: Alright. 

Personally I'd still ignore any indication/notification that contains invalid data (e.g. is_last different from 0 or 1), but the HAP spec does not specify that it shall be ignored directly, and no test case seemingly exist for it (yet), so won't block this PR based on that :)

File: subsys/bluetooth/audio/has_client.c
Comment (Thalley) [reply subsys/bluetooth/audio/has_client.c]: Had to read up on it in the spec :) Slightly confusing operation, but I get it now :)

File: include/zephyr/bluetooth/audio/has.h
Comment (Thalley) [inline include/zephyr/bluetooth/audio/has.h:76]: Should these be past tense? 

e.g. if the type is `BT_HAS_PRESET_DELETE`, doesn't that imply that the preset was _delete*d*_?

File: include/zephyr/bluetooth/audio/has.h
Comment (Thalley) [inline include/zephyr/bluetooth/audio/has.h:88]: Wouldn't mind if the fields in these structs had doxygen comments/documentation :)

File: include/zephyr/bluetooth/audio/has.h
Comment (Thalley) [inline include/zephyr/bluetooth/audio/has.h:132]: Is it `NULL` in remaining cases? If so, please add that in the documentation

File: subsys/bluetooth/audio/has_client.c
Comment (Thalley) [inline subsys/bluetooth/audio/has_client.c:27]: Can we reuse the same flag? 

I guess that regardless of a discover or a CP operation is in progress, we shouldn't allow any other operation?

File: subsys/bluetooth/audio/has_client.c
Comment (Thalley) [reply subsys/bluetooth/audio/has_client.c:113]: The same question goes for the name length: If the buf->len is e.g. 60, doesn't the imply an invalid preset?

File: subsys/bluetooth/audio/has_client.c
Comment (Thalley) [inline subsys/bluetooth/audio/has_client.c:110]: Given that `preset_event` has an error field, should we call the callback here?

Comment (Thalley) [reply subsys/bluetooth/audio/has_client.c:110]: Applies multiple places

File: subsys/bluetooth/audio/has_client.c
Comment (Thalley) [inline subsys/bluetooth/audio/has_client.c:149]: Given that the generic update does not necessarily mean that a preset was removed (in fact the spec states that it `The Generic Update ChangeId shall be used to inform clients when a new preset record is added to the list of preset records`) we shouldn't call the `preset_event` with `BT_HAS_PRESET_DELETE` unconditionally. 

Can we check if there is a deletion by checking the cache?

File: subsys/bluetooth/audio/has_client.c
Comment (Thalley) [inline subsys/bluetooth/audio/has_client.c:204]: Instead of doing `inst = inst_by_conn(conn);` in each handler function, we should do that in the caller and provide a `struct has_inst *inst` instead of `struct bt_conn *conn`

File: subsys/bluetooth/audio/has_client.c
Comment (Thalley) [inline subsys/bluetooth/audio/has_client.c:900]: Prefer to use `CHECKIF` for values that the caller should know are valid/invalid

```suggestion
	CHECKIF(start_index == BT_HAS_PRESET_INDEX_NONE) {
		return -EINVAL;
	}

	CHECKIF(count == 0u) {
		return -EINVAL;
	}
```

File: subsys/bluetooth/shell/has_client.c
Comment (Thalley) [inline subsys/bluetooth/shell/has_client.c:149]: Please add help documentation for the arguments

**Review by Thalley — CHANGES_REQUESTED** [review]

File: subsys/bluetooth/audio/has_client.c
Comment (MariuszSkamra) [reply subsys/bluetooth/audio/has_client.c:149]: The question is where shall be the cache, what should be it's size and how the cache should look like (should we cache every preset or the application should decide which presets it wants to cache).

The problem is this is implementation specific :(

If you have an idea how it should look like I would be happy to hear because I do not know where to put it and how to handle preset changes. I favor an option that the application takes care of that

File: include/zephyr/bluetooth/audio/has.h
Comment (MariuszSkamra) [reply include/zephyr/bluetooth/audio/has.h:76]: Could be

File: subsys/bluetooth/audio/has_client.c
Comment (MariuszSkamra) [reply subsys/bluetooth/audio/has_client.c:27]: Not exactly, as I use `HAS_DISCOVER_IN_PROGRESS` in one, additional specific check

File: subsys/bluetooth/audio/has_client.c
Comment (Thalley) [reply subsys/bluetooth/audio/has_client.c:149]: Since the HAS client have been implemented with the idea that the data lives in the application, then perhaps we should simply just forward the notifications/indications directly to the application (with verification of the values of course), instead of having the `preset_event`? 

If the data and logic exists in the application layer, then I guess the stack shouldn't try to be smart and simply act as a layer that does minimum verification (and of course the translations from `bt_conn` to `bt_has`).

Comment (MariuszSkamra) [reply subsys/bluetooth/audio/has_client.c:149]: OK, done

File: include/zephyr/bluetooth/audio/has.h
Comment (Thalley) [inline include/zephyr/bluetooth/audio/has.h:74]: Could use some documentation as well :)

Comment (Thalley) [inline include/zephyr/bluetooth/audio/has.h:74]: Could use some documentation as well :)

File: include/zephyr/bluetooth/audio/has.h
Comment (Thalley) [inline include/zephyr/bluetooth/audio/has.h:117]: Would it make sense that instead of supplying the `is_last` to the application, that we somehow collect and aggregate all the responses in an array and provide a single callback? 
I don't suppose the client can in any way stop the "read" (which is, IIRC, a multiple of notifications, right?)?

Comment (Thalley) [inline include/zephyr/bluetooth/audio/has.h:117]: Would it make sense that instead of supplying the `is_last` to the application, that we somehow collect and aggregate all the responses in an array and provide a single callback? 
I don't suppose the client can in any way stop the "read" (which is, IIRC, a multiple of notifications, right?)?

File: include/zephyr/bluetooth/audio/has.h
Comment (Thalley) [inline include/zephyr/bluetooth/audio/has.h:196]:
```suggestion
 * The preset records are returned in the @ref bt_has_client_cb.preset_read_rsp callback (called once for each preset).
```

Comment (Thalley) [inline include/zephyr/bluetooth/audio/has.h:196]:
```suggestion
 * The preset records are returned in the @ref bt_has_client_cb.preset_read_rsp callback (called once for each preset).
```

**Review by Thalley — CHANGES_REQUESTED** [review]

File: subsys/bluetooth/audio/has_client.c
Comment (Thalley) [inline subsys/bluetooth/audio/has_client.c:97]: Is there any advantage of having this struct internally and expose 
```c
struct bt_has_preset_record {
	uint8_t index;
	enum bt_has_properties properties;
	const char name[];
};
```
externally? 

Could we not use this struct in the public API as well?

File: subsys/bluetooth/audio/has_client.c
Comment (Thalley) [inline subsys/bluetooth/audio/has_client.c:117]: Please use a `#define` instead of magic numbers for the `is_last` value check :)

**Review by Thalley — CHANGES_REQUESTED** [review]

File: subsys/bluetooth/audio/has_client.c
Comment (MariuszSkamra) [reply subsys/bluetooth/audio/has_client.c:97]: changed that

File: include/zephyr/bluetooth/audio/has.h
Comment (MariuszSkamra) [reply include/zephyr/bluetooth/audio/has.h:117]: I guess that would require having additional buffer for this that might increase the memory footprint unnecesarily. I don't feel this is needed

File: subsys/bluetooth/audio/has_client.c
Comment (MariuszSkamra) [reply subsys/bluetooth/audio/has_client.c:110]: I don't feel this information could be useful

File: include/zephyr/bluetooth/audio/has.h
Comment (Thalley) [reply include/zephyr/bluetooth/audio/has.h:117]: Agreed

**Review by Thalley — DISMISSED** [review]

**Comment by Thalley** [comment]
> Added callback guards

Can you elaborate? I don't see any callback guards :s

File: subsys/bluetooth/audio/has_client.c
Comment (MariuszSkamra) [inline subsys/bluetooth/audio/has_client.c:216]: @Thalley Ah, sorry. Here ^^ So the notification is not handled if the callback is not set

**Review by Thalley — DISMISSED** [review]

**Review by asbjornsabo — CHANGES_REQUESTED** [review]
Some comments.

In addition I am wondering about this possibility to enqueue multiple control point operations.
- is it required
- can it lead to operations arriving out-of-order?  (EATT bearers are not sycnhronised, I believe.)

File: subsys/bluetooth/shell/has_client.c
Comment (MariuszSkamra) [reply subsys/bluetooth/shell/has_client.c:33]: If failed the `has` will be NULL, so the `inst` will be set to NULL as well

File: include/zephyr/bluetooth/audio/has.h
Comment (MariuszSkamra) [reply include/zephyr/bluetooth/audio/has.h:114]: I've changed it to `The callback is called when the preset read response is sent by the remote server.`

File: include/zephyr/bluetooth/audio/has.h
Comment (MariuszSkamra) [reply include/zephyr/bluetooth/audio/has.h:148]: Not really, the Server might send multiple Preset Changed notifications like:
`Generic Update` followed by `Preset Available` etc. and the last one has to have `is_last` flag set, so that the preset list update operation can be considered concluded

File: include/zephyr/bluetooth/audio/has.h
Comment (MariuszSkamra) [reply include/zephyr/bluetooth/audio/has.h:161]: See my previous comment

**Comment by MariuszSkamra** [comment]
> Some comments.
> 
> In addition I am wondering about this possibility to enqueue multiple control point operations.
> 
> * is it required
> * can it lead to operations arriving out-of-order?  (EATT bearers are not sycnhronised, I believe.)

In one of the previous versions of this PR, I had a pool for control point requests see my comment https://github.com/zephyrproject-rtos/zephyr/pull/45071#discussion_r858461582. But decided to remove it for now, so that only one operation is allowed at a time (just like for other LE Audio related client-side implementations)

File: include/zephyr/bluetooth/audio/has.h
Comment (asbjornsabo) [reply include/zephyr/bluetooth/audio/has.h:148]: Still not clear to me.
There are three callbacks, `preset_update`, `preset_deleted` and `preset_availability`.  

The `preset_deleted` is for "preset deletion notifications", but the `is_last` parameter refers to the "update operation".
Similar for the `preset_availability`.

Does the "update" operation have three possible callbacks, of which only one is called update?
And are preset deletion and preset availability part of the preset update?

Comment (MariuszSkamra) [reply include/zephyr/bluetooth/audio/has.h:148]: Then multiple of the preset records changed on the server the server sends the series of notifications that carry information about *updates in the list* of its internal records.

The notification is called *Preset Changed* notification and one of it's parameters is *ChangeID*. The *ChangeID* determines which callback will be called on our side. Once the last *Preset Changed* notification is received, appropriate callback is called (whether the last one *ChangeID* was Preset Deleted, Generic Update, Preset Available or Preset Unavailable) with the `is_last` flag set.

If that is still confusing I can make it as a single callback (it was done like this initially, see comment https://github.com/zephyrproject-rtos/zephyr/pull/45071#discussion_r858654067 but we agreed to split it to multiple callbacks)

**Review by asbjornsabo — APPROVED** [review]

File: subsys/bluetooth/audio/has_client.c
Comment (Thalley) [inline subsys/bluetooth/audio/has_client.c:164]: Suggestion for improved readability: 

```suggestion
	uint8_t index;
	
	if (buf->len < sizeof(index {
		BT_ERR("malformed PDU");
		return;
	}
	
	index = net_buf_simple_pull_u8(buf);

	client_cb->preset_deleted(&inst->has, index, is_last);
```

(applies multiple places).

File: subsys/bluetooth/audio/has_client.c
Comment (Thalley) [reply subsys/bluetooth/audio/has_client.c:216]: Would it make more sense to move the `if (client_cb->preset_availability) {` to the `handle_preset_availability` function? Keeping this `switch` more "clean" and also avoiding having `if (client_cb->preset_availability) {` twice her?

**Review by Thalley — APPROVED** [review]
LGTM. 

Two minor suggestions that can be done, but they are minor enough that this can be merged without it.

**Comment by mbolivar-nordic** [comment]
> Two minor suggestions that can be done, but they are minor enough that this can be merged without it.

Merging without them since today is code freeze

**Pull request merged**
